attributable_error_crypto.go

```diff
@@ -0,0 +1,103 @@
+package sphinx
+
+import (
+	"crypto/hmac"
+	"crypto/sha256"
+)
+
+type payloadSource byte
+
+const (
+	// payloadIntermediateNode is a marker to signal that this attributable
+	// error payload is originating from a node between the payer and the error
+	// source.
+	payloadIntermediateNode payloadSource = 0
+
+	// payloadErrorNode is a marker to signal that this attributable error
+	// payload is originating from the error source.
+	payloadErrorNode payloadSource = 1
+)
+
+type AttributableErrorStructure struct {
+	HopCount        int
```

Comment:
these are `u16`,`u8` etc in the BOLT, maybe use the same here?

Reply:
I used ints here to prevent casting to a wider data type in the various calculations that use these values. I think you could argue that the `byte` that is specified in the bolt is a limitation on the wire level and that the code in this repo doesn't need to follow that limitation necessarily?

Reply:
I think it would be nice to have the types 1:1. How bad would the casting be?

Reply:
I think the casting overhead would be negligible in terms of affected code sites. Inheriting the proper int types here aides in readability IMO. 

Also in most cases, you don't really need to cast for comparisons, as you can make generic funcs that use `constraints. Integer` or w/e.

Reply:
Also missing `godoc` comments here. 

Is this an error struct (actually `error` type), or something else entirely? In either case, adding "Structure" to the end seems superflous.

Reply:
It is not an error struct, but a description of the actual structure of the onion failure. I think the addition `Structure` is appropriate in this case. Added godoc to explain.